engine/engine_test.go

```diff
@@ -347,6 +347,242 @@ func TestEngine_NotifyPutThenNotifyRemove(t *testing.T) {
 	require.NotEqual(t, gotLatestAfterRmAdCid, gotLatestAdCid)
 }
 
+func TestEngine_WhenPublishForMultipleProviders_AdsShouldBeOnTheSameChain(t *testing.T) {
+	ctx := contextWithTimeout(t)
+	rng := rand.New(rand.NewSource(1413))
+
+	mhs1 := testutil.RandomMultihashes(t, rng, 42)
+	mhs2 := testutil.RandomMultihashes(t, rng, 42)
+
+	provider1id := "Provider 1"
+	provider1Addrs := []string{"/ip4/0.0.0.0/tcp/1234/http"}
+	provider2id := "Provider 2"
+	provider2Addrs := []string{"/ip4/0.0.0.0/tcp/4321/http"}
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID1 := []byte("fish")
+	wantContextID2 := []byte("bird")
+	subject.RegisterMultihashLister(func(ctx context.Context, p string, contextID []byte) (provider.MultihashIterator, error) {
+
+		if string(contextID) == string(wantContextID1) {
+			return provider.SliceMultihashIterator(mhs1), nil
+		} else if string(contextID) == string(wantContextID2) {
+			return provider.SliceMultihashIterator(mhs2), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, provider1id, provider1Addrs, wantContextID1, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+
+	gotLatestAdCid, ad, err := subject.GetLatestAdv(ctx)
+	require.NoError(t, err)
+	require.Equal(t, gotLatestAdCid, gotPutAdCid1)
+	require.Equal(t, ad.Provider, provider1id)
+	require.Equal(t, ad.Addresses, provider1Addrs)
+
+	gotPutAdCid2, err := subject.NotifyPut(ctx, provider2id, provider2Addrs, wantContextID2, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid2)
+
+	gotLatestAdCid, ad, err = subject.GetLatestAdv(ctx)
+	require.NoError(t, err)
+	require.Equal(t, gotLatestAdCid, gotPutAdCid2)
+	require.Equal(t, ad.Provider, provider2id)
+	require.Equal(t, ad.Addresses, provider2Addrs)
+	require.Equal(t, ad.PreviousID.(cidlink.Link).Cid, gotPutAdCid1)
+}
+
+func TestEngine_WhenProviderIsEmpty_ShouldDefaultToTheProviderFromConfiguration(t *testing.T) {
```

Comment:
`TestEngine_UsesDefaultProviderWhenNoneGiven`

Totally cosmetic suggestion. See if you can find shorter names for the remaining tests.

Reply:
fixed